### PR TITLE
[MERGE] web: Re-implement the XLS grouped export

### PR DESCRIPTION
**Pull request by LucasLefevre**
Purpose
=======

The commit 2849b5c introduces a new export mechanism of grouped
list views to xls files. 

The issue with this development is mainly that the displayed records
are exported, instead of all the records that match the search
parameters.

To export all the records we cannot rely on the data from the web
client. This implies to revert 2849b5c, and implement it in a better
way.

Functional Spec
===============

Add the support of grouped exports.
Allow the user to export all records in one click from
the listview, without having to go through the export modal,
taking into account the domain, groupbys, and visible fields.

Technical Spec
==============

When exporting (whether it is from the modal or from the shortcut),
any groupby(s) set on the listview should be taken into account (all unfolded)
- UNLESS the export is import-compatible
- each subgroup header has an indentation compared to its parent
- the 'group headers' in the exported file should contain the
  same info (label, field aggregates) as it has in the listview.

New secondary button on the tree view with label 'EXPORT' (to be confirmed...)
- the export shortcut disregard the selected records, it exports all records
  according to the domain.
- the button is visible even if there is no selected records.
- essentially the export shortcut does the same thing as the following:
    - select all records
    - hit 'action' then 'export'
    - hit 'export'

Define a boolean attribute on <tree> to specify whether or not the export
shortcut should be displayed

Task 2072910

### CODE REVIEW

**Comment by VincentSchippefilt** [comment]
@aab-odoo could you check the JS part pls ?

File: addons/web/static/src/js/views/list/list_controller.js [hunk 1]
```diff
@@ -109,7 +54,7 @@ var ListController = BasicController.extend({
         this.selectedRecords = params.selectedRecords || [];
         this.multipleRecordsSavingPromise = null;
         this.activeActions = _.extend({}, this.activeActions, {
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/list/list_controller.js]: this should be done in ListView

File: addons/web/static/src/js/views/list/list_controller.js [hunk 2]
```diff
@@ -400,29 +342,13 @@ var ListController = BasicController.extend({
         });
     },
     /**
-     * Export the current list data in a xls file.
-     *
+     * Return the export dialog widget
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/list/list_controller.js]: missing `@return` then

File: addons/web/static/src/js/views/list/list_controller.js [hunk 3]
```diff
@@ -400,29 +342,13 @@ var ListController = BasicController.extend({
...
-        groups = processGroups(groups, columns.map(c => c.field), allFields)
-
-        return session.get_file({
-            url: '/web/list/export_xls',
-            data: {data: JSON.stringify({columns, groups, title: this._title})},
-            complete: framework.unblockUI,
-            error: (error) => this.call('crash_manager', 'rpc_error', error),
-        });
+    _getExportDialogWidget() {
+        var record = this.model.get(this.handle);
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/list/list_controller.js]: it is not a record, i think.. So i'd name it `state` or something

File: addons/web/static/src/js/widgets/data_export.js [hunk 4]
```diff
@@ -49,6 +49,11 @@ var DataExport = Dialog.extend({
         this.exports = new data.DataSetSearch(this, 'ir.exports', this.record.getContext());
         this.rowIndex = 0;
         this.rowIndexLevel = 0;
+        this.isCompatibleMode = false;
+        this.groupby = this.getParent().renderer.state.groupedBy;
+        let domain = this.getParent().getActiveDomain();
+        this.domain = domain || this.record.domain;
+        this.idsToExport = domain ? false: this.getParent().getSelectedIds();
```
Comment (aab-odoo) [inline addons/web/static/src/js/widgets/data_export.js]: my eyes are bleeding... The required information should be given at initialization

Comment (LucasLefevre) [reply addons/web/static/src/js/widgets/data_export.js]: This code was only moved  from `start` to `init` (and re-written - was even more eye-bleeding before). I didn't want to change too much. But ok to take this opportunity to clean this.
Done

**Comment by LucasLefevre** [comment]
@aab-odoo Thanks for your review!
Your comments have been applied.
BTW, I added JS tests if you want to take a look.

File: addons/web/controllers/main.py [hunk 5]
```diff
@@ -607,6 +607,159 @@ def get_qweb_templates(cls, addons, db=None, debug=False):
...
+        self.write(row, column, cell_value, cell_style)
+
+class GroupExportXlsxWriter(ExportXlsxWriter):
+
+    def __init__(self, fields, row_count=0):
+        self.fields = fields
+        super().__init__([f['label'].strip() for f in self.fields], row_count)
+
+    def write_group(self, row, column, group_name, group, group_depth=0):
+        group_name = group_name if isinstance(group_name, str) else group_name and group_name[1] or odoo._("Undefined")
```
Comment (RomainLibert) [inline addons/web/controllers/main.py]:
```suggestion
        group_name = group_name if isinstance(group_name, str) else group_name and group_name[1] or _("Undefined")
```

As `_` is already imported, see line 41

**Comment by tivisse** [comment]
@xmo-odoo Do you wish to take a look at this ?

File: addons/web/controllers/main.py [hunk 6]
```diff
@@ -1602,26 +1755,43 @@ def from_data(self, fields, rows):
...
+            # read_group(lazy=False) returns a dict only for final groups (with actual data),
+            # not for intermediary groups. The full group tree must be re-constructed.
+            tree = GroupsTreeNode(Model, field_names, groupby)
+            for leaf in groups_data:
+                tree.insert_leaf(leaf)
+
+            response_data = self.from_group_data(fields, tree)
+        else:
+            Model = Model.with_context(import_compat=import_compat)
+            records = Model.browse(ids) or Model.search(domain, offset=0, limit=False, order=False)
```
Comment (RomainLibert) [inline addons/web/controllers/main.py]: This is probably better performance wise
```suggestion
            records = Model.browse(ids) if ids else Model.search(domain, offset=0, limit=False, order=False)
```

File: addons/web/controllers/main.py [hunk 7]
```diff
@@ -1602,26 +1755,43 @@ def from_data(self, fields, rows):
...
-
         if import_compat:
             columns_headers = field_names
         else:
             columns_headers = [val['label'].strip() for val in fields]
 
-        return request.make_response(self.from_data(columns_headers, import_data),
+        Model = request.env[model].with_context(**params.get('context', {}))
+        groupby = params.get('groupby')
+        if not import_compat and groupby:
```
Comment (RomainLibert) [inline addons/web/controllers/main.py]: Does this mean that if you give `import_compat` you cannot export using the group by?

File: addons/web/controllers/main.py [hunk 8]
```diff
@@ -1602,26 +1755,43 @@ def from_data(self, fields, rows):
...
+
+            response_data = self.from_group_data(fields, tree)
+        else:
+            Model = Model.with_context(import_compat=import_compat)
+            records = Model.browse(ids) or Model.search(domain, offset=0, limit=False, order=False)
+
+            if not Model._is_an_ordinary_table():
+                fields = [field for field in fields if field['name'] != 'id']
+
+            import_data = records.export_data(field_names).get('datas',[])
```
Comment (RomainLibert) [inline addons/web/controllers/main.py]:
```suggestion
            export_data = records.export_data(field_names).get('datas',[])
```

File: addons/web/controllers/main.py [hunk 9]
```diff
@@ -1602,26 +1755,43 @@ def from_data(self, fields, rows):
...
+            response_data = self.from_data(columns_headers, import_data)
```
Comment (RomainLibert) [inline addons/web/controllers/main.py]:
```suggestion
            response_data = self.from_data(columns_headers, export_data)
```

File: addons/web/controllers/main.py (hunk 7 above)
Comment (LucasLefevre) [reply addons/web/controllers/main.py]: yep, as per spec import compatibility is stronger than groupby

**Comment by tivisse** [comment]
@robodoo r+ merge

**Comment by tivisse** [comment]
@robodoo r+

File: addons/web/controllers/main.py (hunk 5 above)
Comment (sswapnesh) [reply addons/web/controllers/main.py]: @RomainLibert  Would like to have your review on https://github.com/odoo/odoo/pull/38007 as I am touching this line

**Comment by Dimocave** [comment]
Hi, this is really great, thanks! I suggest a change for exporting numbers as numbers and not as text:
You have to replace both `records.export_data(self._fields).get('datas',[])` for `records.export_data(self._fields,  raw_data=True).get('datas',[])` and add self.float_style = self.workbook.add_format().set_num_format('#,##0.00')
        in  
`class ExportXlsxWriter:

    def __init__(self, field_names, row_count=0):`

and at last: add this:
`elif isinstance(cell_value, float):
            cell_style = self.float_style`
in `def write_cell(self, row, column, cell_value):`

Sorry if I don't do the commit as it should be, I don't have the same version of odoo and it is a bit tricky for me...

**Pull request merged**